### Python Shallow Copy and Deep Copy

In [1]:
# Copy an Object in Python

Example 1: Copy using = operator

In [3]:
old_list = [[1, 2, 3], [4, 5, 6], [7, 8, 'a']]
new_list = old_list

new_list[2][2] = 9

print('Old List:', old_list)
print('ID of Old List:', id(old_list))

print('-'*30)
print('NEw List:', old_list)
print('ID of New List:', id(new_list))

Old List: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
ID of Old List: 2260693085760
------------------------------
NEw List: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
ID of New List: 2260693085760


As you can see from the output both variables old_list and new_list shares the same id i.e 2260693085760.

So, if you want to modify any values in new_list or old_list, the change is visible in both.

In [4]:

old_list[1][1] = 9

print('Old List:', old_list)
print('ID of Old List:', id(old_list))

print('-'*30)
print('NEw List:', old_list)
print('ID of New List:', id(new_list))

Old List: [[1, 2, 3], [4, 9, 6], [7, 8, 9]]
ID of Old List: 2260693085760
------------------------------
NEw List: [[1, 2, 3], [4, 9, 6], [7, 8, 9]]
ID of New List: 2260693085760


Sometimes you may want to have the original values unchanged and only modify the new values or vice versa.
In Python, there are two ways to create copies:

1. Shallow Copy

2. Deep Copy

### Copy Module

Example 2: Create a copy using shallow copy

In [5]:
import copy

old_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
new_list = copy.copy(old_list)

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
New list: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]


### Adding [4, 4, 4] to old_list, using shallow copy

In [6]:
import copy

old_list = [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
new_list = copy.copy(old_list)

old_list.append([4, 4, 4])

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 1, 1], [2, 2, 2], [3, 3, 3], [4, 4, 4]]
New list: [[1, 1, 1], [2, 2, 2], [3, 3, 3]]


### Adding new nested object using Shallow copy

In [7]:
import copy

old_list = [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
new_list = copy.copy(old_list)

old_list[1][1] = 'AA'

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 1, 1], [2, 'AA', 2], [3, 3, 3]]
New list: [[1, 1, 1], [2, 'AA', 2], [3, 3, 3]]


### Deep Copy

### Copying a list using deepcopy()

In [8]:
import copy

old_list = [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
new_list = copy.deepcopy(old_list)

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
New list: [[1, 1, 1], [2, 2, 2], [3, 3, 3]]


Adding a new nested object in the list using Deep copy

In [10]:
import copy

old_list = [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
new_list = copy.deepcopy(old_list)

old_list[1][0] = 'BB'

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 1, 1], ['BB', 2, 2], [3, 3, 3]]
New list: [[1, 1, 1], [2, 2, 2], [3, 3, 3]]


In [11]:
import copy

old_list = [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
new_list = copy.deepcopy(old_list)

new_list[1][0] = 'AAA'

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
New list: [[1, 1, 1], ['AAA', 2, 2], [3, 3, 3]]
